### PR TITLE
[babel 8] Remove support for the `2018-09` decorators proposal

### LINKED ISSUE #8750: Error with new Decorator Plugin and HTMLElement instance with new
**Current Behavior**

This class extend from HTMLImport run perfectly (it's very simple):

```js
class Counter extends HTMLElement {
  constructor() {
    super();
  }
}

window.customElements.define('my-counter', Counter);
const counter = new Counter();
```

Now, I use @littledan example decorator:

```js
// Define the class as a custom element with the given tag name
function defineElement(tagName) {
  // In order for a decorator to take an argument, it takes that argument
  // in the outer function and returns a different function that's called
  // when actually decorating the class (manual currying).
  return function(classDescriptor) {
    let { kind, elements } = classDescriptor;
    console.assert(kind == "class");
    return {
      kind,
      elements,
      // This callback is called once the class is otherwise fully defined
      finisher(klass) {
        window.customElements.define(tagName, klass);
      }
    };
  };
}

@defineElement('my-counter')
class Counter extends HTMLElement {
  constructor() {
    super();
  }
}

const counter = new Counter();
```

When that code is transformed with the new Decorator plugin (thank @nicolo-ribaudo) and run it on a browser, this error is displayed:

```
Uncaught TypeError: Failed to construct 'HTMLElement': Please use the 'new' operator, this DOM object constructor cannot be called as a function.
    at new Counter (<anonymous>:100:73)
    at <anonymous>:116:15
```

**Expected behavior/code**

The decoration don't change the Custom Element behavior and must be run without problems.

**Babel Configuration (.babelrc.js)**

```js
const presets = [ "@babel/preset-env" ];
const plugins = [
  [ "@babel/plugin-proposal-decorators", { "legacy": false, decoratorsBeforeExport: false } ]
];
module.exports = { presets, plugins };
```

**Environment**
- Babel version(s): 7.1.0

### LINKED ISSUE #9220: [Bug]: Decorators Conflict With Function Overloading
## Bug Report

**Current Behavior**

When using **@babel/plugin-proposal-decorators** alongside **@babel/preset-typescript** [function overloading](https://www.typescriptlang.org/docs/handbook/functions.html#overloads) does not work.

`TypeError: Duplicated element (nameOfOverloadedClassMethod)`

This is the transpiled output:

```javascript
{
  kind: "field",
  key: "nameOfOverloadedClassMethod",
  value: void 0
}, {
  kind: "field",
  key: "nameOfOverloadedClassMethod",
  value: void 0
}, {
  kind: "method",
  key: "nameOfOverloadedClassMethod",
  value: function value(_bar) {
    // actual body of method
  }
}
```

**Input Code**

```typescript
class Foo {
  nameOfOverloadedClassMethod(bar: string);
  nameOfOverloadedClassMethod(bar: number);
  nameOfOverloadedClassMethod(bar) {
    // some implementation
  }
}
```

**Expected behavior/code**

The overloaded methods are removed when transpiling before decorators mangle the class.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  loader: 'babel-loader',
  options: {
    presets: [
      [
        '@babel/preset-env',
        {
          modules: false,
          useBuiltIns: 'usage'
        }
      ],
      '@babel/typescript'
    ],
    plugins: [
      ['@babel/plugin-proposal-decorators', {
        decoratorsBeforeExport: true
      }],
      '@babel/plugin-proposal-class-properties',
      '@babel/plugin-proposal-object-rest-spread',
      '@babel/plugin-transform-runtime',
      'angularjs-annotate'
    ]
  }
}
```

**Environment**
- Babel version(s): v7.2.0
- Node/npm version: Node 9.11.2
- OS: OSX 10.14.1
- Monorepo: no
- How you are using Babel: `loader`

### LINKED ISSUE #9454: @babel/plugin-proposal-decorators not working with @babel/plugin-transform-runtime
## Bug Report

**Current Behavior**
Decorator helpers are not imported if `@babel/plugin-transform-runtime` is enabled, instead, they are still defined in file as if no transform-runtime plugin is used. 

**Input Code**
- REPL or Repo link if applicable:
https://runkit.com/lodin/decorators-transform-runtime-issue

```js
@cls
class A {
  @prop
  test() {}
}
```

**Expected behavior/code**
All helper functions in the file are replaced with imports from `@babel/runtime`. 

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  plugins: [
    [require("@babel/plugin-proposal-decorators"), {decoratorsBeforeExport: true}],
    require("@babel/plugin-proposal-class-properties"),
    [require("@babel/plugin-transform-runtime"), {
      "corejs": false,
      "helpers": true,
      "regenerator": false,
      "useESModules": true
    }],
  ],
  presets: [],
}
```

**Environment**
- Babel version(s): 7.2.2
- Node/npm version: node 11.8.0/npm 6.5.0
- OS: Windows 10
- Monorepo: [e.g. yes/no/Lerna]: no
- How you are using Babel: [e.g. `cli`, `register`, `loader`]: cli

**Additional context/Screenshots**
Add any other context about the problem here. If applicable, add screenshots to help explain.

I can find `@babel/runtime/helpers/esm/decorate.js` file in `node_modules`, so only import replacing doesn't work for some reason.

### LINKED ISSUE #9602: Decorators and babel polyfills
## Bug Report
It seems that the current babel implementation of [decorators](https://github.com/tc39/proposal-decorators) relies on using Symbol. However, even with babel polyfills enabled, in this instance Symbol is not polyfilled and an error is thrown in IE11.

**Current Behavior**
IE11 throws an error when using babelled decorators: `'Symbol' is undefined`

**Input Code**
Using a decorator anywhere in the code

**Expected behavior/code**
The Symbol polyfill perhaps should be imported automatically

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "presets": [
    ["@babel/preset-env", {
      "targets": {
        "browsers": [
          "ie 11",
          "last 2 versions",
          "> 1%"
        ]
      },
      "modules": false,
      "useBuiltIns": "usage"
    }]
  ],
  "plugins": [
    ["@babel/plugin-proposal-class-properties"],
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": false }],
    ....
```

**Environment**
* Browser: IE11 (and probably any IE)
* Babel version(s): v7.2.0
* Node/npm version: node 10.3.0/npm 6.1.0
* OS: OSX 10.14.3
* Monorepo: ?
* How you are using Babel: ?

**Possible Solution**
Currently you can get round it by simply mentioning `Symbol` at the top of your project

### LINKED ISSUE #9999: Decorators plugin throws error in ie 11
## Bug Report

**Current Behavior**
ie 11 fully fails with `throw new TypeError("@@toPrimitive must return a primitive value.")}`

**Input Code**
```js
@customElement('test')
class TestApp extends connect(store)(LitElement)
{
  @property({ type: String }) _Test = 'Hello';
```

**Expected behavior/code**
It's meant to not throw an error.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "browserslist": [
    ">= 0.2%",
    "not dead",
    "not ie <= 10",
    "not op_mini all",
    "ie 11",
    "chrome 41"
  ],

  "use strict";

const {minify} = require('terser');

module.exports = {
  presets: [
    ['@babel/env', {
      useBuiltIns: 'usage',
      loose:       true,
      modules:     'amd',
      corejs:      3
    }]
  ],
  plugins: [
    ['template-html-minifier', {
      modules: {
        'lit-html': ['html', 'svg'], // lit-html
        'lit-element': ['html', 'svg'], // LitElement
        '@polymer/polymer/polymer-element': ['html'], // Polymer 3 - exported from Polymer Element
        '@polymer/polymer/lib/utils/html-tag.js': ['html'], // Polymer 3 - exported from utils (used by PolymerElements family)
      },
      htmlMinifier: {
        caseSensitive:      true,
        collapseWhitespace: true,
        removeComments:     true,
        sortAttributes:     false,
        sortClassName:      false,
        minifyCSS:          true,
        minifyJS:           (code) => minify(code).code,
        useShortDoctype:    true
      },
    }],
    ['@babel/transform-runtime', {
      sourceType: 'unambiguous',
      corejs: 3
    }],
    ['@babel/proposal-decorators', {
      decoratorsBeforeExport: true,
    }],
    ['@babel/proposal-class-properties', {
      loose: false
    }],
    ['@babel/proposal-object-rest-spread'],
    ['@babel/syntax-dynamic-import']
  ]
};

"devDependencies": {
    "@babel/core": "7.4.4",
    "@babel/plugin-proposal-class-properties": "7.4.4",
    "@babel/plugin-proposal-decorators": "7.4.4",
    "@babel/plugin-proposal-object-rest-spread": "7.4.4",
    "@babel/plugin-syntax-dynamic-import": "7.2.0",
    "@babel/plugin-transform-runtime": "7.4.4",
    "@babel/preset-env": "7.4.4",
    "acorn": "6.1.1",
    "babel-eslint": "11.0.0-beta.0",
    "babel-loader": "8.0.6",
    "babel-loader-exclude-node-modules-except": "1.0.1",
    "babel-plugin-template-html-minifier": "3.0.0",

cli is just webpack + webpack dev server using the configs.
}
```

**Environment**
- Babel version(s): 7 4 4
- Node/npm version: 12.2
- OS: win7
- Monorepo: yes
- How you are using Babel: loader

### LINKED ISSUE #12591: Decorator helpers use find, which breaks ie11
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

When using decorators through `@babel/plugin-proposal-decorators` and `@babel/plugin-transform-runtime`, [code is included that uses find](https://github.com/babel/babel/blob/a967910b8b9b41e0abac3ce622b0808ecd36ca0f/packages/babel-helpers/src/helpers.js#L2020), which is [unsupported in ie11](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#browser_compatibility). 

**Input Code**

```js
@decorator
class MyClass {}

function decorator() {}
```

**Expected behavior**

I was under the impression that babel would output code that works with what I have in my `.browserslistrc` file, which includes ie11. I have `@babel/preset-env` set up with `useBuiltIns: "usage"` and core-js@3, but I am building through webpack and ignore node_modules. Should I allow babel to transform itself?

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
module.exports = {
  plugins: [
    ['@babel/plugin-proposal-decorators', { decoratorsBeforeExport: true }],
    '@babel/plugin-proposal-class-properties',
    [
      '@babel/plugin-transform-runtime',
      {
        useESModules: true,
        version: require('@babel/runtime/package.json').version,
      },
    ],
  ],
  presets: [
    ['@babel/preset-env', { useBuiltIns: 'usage', corejs: 3 }],
    '@babel/preset-typescript',
  ],
};
```

**Environment**

```
  System:
    OS: macOS Mojave 10.14.6
  Binaries:
    Node: 12.20.0 - ~/.nvm/versions/node/v12.20.0/bin/node
    Yarn: 1.21.1 - ~/.yarn/bin/yarn
    npm: 6.14.10 - ~/.nvm/versions/node/v12.20.0/bin/npm
  npmPackages:
    @babel/core: ^7.12.10 => 7.12.10 
    @babel/plugin-proposal-class-properties: ^7.12.1 => 7.12.1 
    @babel/plugin-syntax-dynamic-import: ^7.2.0 => 7.8.3 
    @babel/preset-env: ^7.12.11 => 7.12.11 
    @babel/preset-typescript: ^7.12.7 => 7.12.7 
    babel-jest: ^26.6.3 => 26.6.3 
    babel-loader: ^8.2.2 => 8.2.2 
    jest: ^26.6.3 => 26.6.3 
    webpack: ^4.43.0 => 4.43.0 
```

**Possible Solution**

Use a helper function that loops instead of calling `.find`.

**Additional context**

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #8750, closes #9454, closes #9220, closes #9602, closes #9999, closes #12591
| Patch: Bug Fix?          |
| Major: Breaking Change?  | Yes, behind the `BABEL_8_BREAKING` flag
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR completely removes support for the second decorators proposal (the one with the big runtime code).
1. We never fully supported the proposal (decorated private elements were left as a todo)
2. That proposal version has been dropped on 2019-03-15, which is almost two years ago (https://github.com/tc39/proposal-decorators/commit/a043d4bfff959648dceb151a1c288bb595ae2b74)
3. After that version (which was the second) there have been other two versions

I'm not removing support for `legacy` decorators because many JS libraries depend on them.

Please review this PR with whitespace diff disabled, since different parts have been  indented one level more (because of the `if (BABEL_8_BREAKING) {` statements).

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12712"></a>

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12712"></a>

### CODE REVIEW

**Review by JLHwung — COMMENTED** [review]
We should also remove `decoratorsBeforeExport` from `@babel/standalone`.

EDIT by @nicolo-ribaudo: the new proposal version still supports this option

**Comment by mbrowne** [comment]
Just curious, when will work begin on an implementation of the latest decorators proposal? Or is it still not stable enough yet to be worth doing that at this point?

**Comment by nicolo-ribaudo** [comment]
I'm in contact with the proposal authors, and it looks like it's becoming stable enough that we can implement it in the next months.

**Comment by wvq** [comment]
For now, Is there any progress on stage 2 ?

**Comment by nicolo-ribaudo** [comment]
We are working on it at https://github.com/babel/babel/pull/14004

**Comment by nicolo-ribaudo** [comment]
CI failure fixed!

**Comment by nicolo-ribaudo** [comment]
Rebased, this is ready for review!

**Comment by nicolo-ribaudo** [comment]
Rebased again

File: packages/babel-plugin-syntax-decorators/src/index.ts [hunk 1]
```diff
@@ -2,39 +2,59 @@ import { declare } from "@babel/helper-plugin-utils";
...
 
+  let { decoratorsBeforeExport } = options;
   if (decoratorsBeforeExport === undefined) {
-    if (version === "2018-09") {
-      throw new Error(
-        "The decorators plugin, when .version is '2018-09' or not specified," +
-          " requires a 'decoratorsBeforeExport' option, whose value must be a boolean.",
-      );
+    if (version === "2021-12") {
+      decoratorsBeforeExport = false;
```
Comment (JLHwung) [inline packages/babel-plugin-syntax-decorators/src/index.ts:49]: Does it mean we can remove the `decoratorsBeforeExport` generator option in Babel 8?

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-syntax-decorators/src/index.ts (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-syntax-decorators/src/index.ts:49]: This PR still keeps the `decoratorsBeforeExport` option but makes it default to `false`, I think I can just remove it given the proposal is now stage 3.

**Comment by nicolo-ribaudo** [comment]
@JLHwung I removed `decoratorsBeforeExport` from the plugin and from generator. I didn't remove it in the parser yet because I first want to make it optional in Babel 7 (so that I don't have to duplicate all the tests between Babel 7 and Babel 8), but it needs to wait for the next minor.

**Review by liuxingbaoyu — APPROVED** [review]

**Pull request merged**